Bitbucket.Authentication/Authentication.cs

```diff
@@ -229,15 +230,34 @@ public void SetCredentials(TargetUri targetUri, Credential credentials, string u
             Trace.WriteLine($"{credentials.Username} at {targetUri.ActualUri.AbsoluteUri}");
 
             // if the url doesn't contain a username then save with an explicit username.
-            if (!targetUri.TargetUriContainsUsername && !string.IsNullOrWhiteSpace(username))
+            if (!targetUri.TargetUriContainsUsername && (!string.IsNullOrWhiteSpace(username)
+                || !string.IsNullOrWhiteSpace(credentials.Username)))
             {
-                Credential tempCredentials = new Credential(username, credentials.Password);
+                var realUsername = GetRealUsername(credentials, username);
+                Credential tempCredentials = new Credential(realUsername, credentials.Password);
                 SetCredentials(targetUri.GetPerUserTargetUri(username), tempCredentials, null);
             }
 
             PersonalAccessTokenStore.WriteCredentials(targetUri, credentials);
         }
 
+        private static string GetRealUsername(Credential credentials, string username)
```

Comment:
While the credentials can\will be stored using the username provided by the user, e.g. email, in the target uri the credentials need to use the 'real' username as returned from Bitbucket.